In [31]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 12: JSAbacusFramework.io ---</h2><p>Santa's Accounting-Elves need help balancing the books after a recent order.  Unfortunately, their accounting software uses a peculiar storage format.  That's where you come in.</p>
<p>They have a <a href="http://json.org/">JSON</a> document which contains a variety of things: arrays (<code>[1,2,3]</code>), objects (<code>{"a":1, "b":2}</code>), numbers, and strings.  Your first job is to simply find all of the <em>numbers</em> throughout the document and add them together.</p>
<p>For example:</p>
<ul>
<li><code>[1,2,3]</code> and <code>{"a":2,"b":4}</code> both have a sum of <code>6</code>.</li>
<li><code>[[[3]]]</code> and <code>{"a":{"b":4},"c":-1}</code> both have a sum of <code>3</code>.</li>
<li><code>{"a":[-1,1]}</code> and <code>[-1,{"a":1}]</code> both have a sum of <code>0</code>.</li>
<li><code>[]</code> and <code>{}</code> both have a sum of <code>0</code>.</li>
</ul>
<p>You will not <span title="Nor are you likely to be eaten by a grue... during *this* puzzle, anyway.">encounter</span> any strings containing numbers.</p>
<p>What is the <em>sum of all numbers</em> in the document?</p>
</article>


In [32]:
tests = [
    {
        "name": "Example 1",
        "s": "[1,2,3]",
        "expected": 6,
    },
    {
        "name": "Example 2",
        "s": '{"a":2,"b":4}',
        "expected": 6,
    },
    {
        "name": "Example 3",
        "s": "[[[3]]]",
        "expected": 3,
    },
    {
        "name": "Example 4",
        "s": '{"a":{"b":4},"c":-1}',
        "expected": 3,
    },
    {
        "name": "Example 5",
        "s": '{"a":[-1,1]}',
        "expected": 0,
    },
    {
        "name": "Example 6",
        "s": '[-1,{"a":1}]',
        "expected": 0,
    },
    {
        "name": "Example 7",
        "s": "[]",
        "expected": 0,
    },
    {
        "name": "Example 8",
        "s": "{}",
        "expected": 0,
    },
]


def sum_all_ints(s: str) -> int:
    return sum(map(int, re.findall(r"-?\d+", s)))


run_tests_params(sum_all_ints, tests)


Test Example 1 passed, for sum_all_ints.
Test Example 2 passed, for sum_all_ints.
Test Example 3 passed, for sum_all_ints.
Test Example 4 passed, for sum_all_ints.
Test Example 5 passed, for sum_all_ints.
Test Example 6 passed, for sum_all_ints.
Test Example 7 passed, for sum_all_ints.
Test Example 8 passed, for sum_all_ints.
Success


In [33]:
with open("../input/day12.txt") as f:
    print(sum_all_ints(f.read()))

156366


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>156366</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Uh oh - the Accounting-Elves have realized that they double-counted everything <em>red</em>.</p>
<p>Ignore any object (and all of its children) which has any property with the value <code>"red"</code>.  Do this only for objects (<code>{...}</code>), not arrays (<code>[...]</code>).</p>
<ul>
<li><code>[1,2,3]</code> still has a sum of <code>6</code>.</li>
<li><code>[1,{"c":"red","b":2},3]</code> now has a sum of <code>4</code>, because the middle object is ignored.</li>
<li><code>{"d":"red","e":[1,2,3,4],"f":5}</code> now has a sum of <code>0</code>, because the entire structure is ignored.</li>
<li><code>[1,"red",5]</code> has a sum of <code>6</code>, because <code>"red"</code> in an array has no effect.</li>
</ul>
</article>

</main>


In [38]:
import json


tests = [
    {
        "name": "Example 1",
        "s": "[1,2,3]",
        "expected": 6,
    },
    {
        "name": "Example 2",
        "s": '[1,{"c":"red","b":2},3]',
        "expected": 4,
    },
    {
        "name": "Example 3",
        "s": '{"d":"red","e":[1,2,3,4],"f":5}',
        "expected": 0,
    },
    {
        "name": "Example 4",
        "s": '[1,"red",5]',
        "expected": 6,
    },
]


def sum_ints_no_reds(s: str) -> int:
    def dfs(token: dict | list | str | int | None) -> int:
        if token is None:
            return 0
        if isinstance(token, str):
            return 0
        if isinstance(token, int):
            return int(token)
        if isinstance(token, list):
            return sum(map(dfs, token), start=0)
        if "red" not in token.values():
            return sum(map(dfs, token.values()), start=0)
        return 0

    return dfs(json.loads(s))


run_tests_params(sum_ints_no_reds, tests)


Test Example 1 passed, for sum_ints_no_reds.
Test Example 2 passed, for sum_ints_no_reds.
Test Example 3 passed, for sum_ints_no_reds.
Test Example 4 passed, for sum_ints_no_reds.
Success


In [39]:
with open("../input/day12.txt") as f:
    print(sum_ints_no_reds(f.read()))

96852


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>96852</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2015">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="12/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22JSAbacusFramework%2Eio%22+%2D+Day+12+%2D+Advent+of+Code+2015&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F12&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22JSAbacusFramework%2Eio%22+%2D+Day+12+%2D+Advent+of+Code+2015+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F12';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
